ff/serving/provider/offline.go

```diff
@@ -89,13 +170,177 @@ func (store *memoryOfflineStore) CreateResourceTable(id ResourceID) (OfflineTabl
 }
 
 func (store *memoryOfflineStore) GetResourceTable(id ResourceID) (OfflineTable, error) {
+	return store.getMemoryResourceTable(id)
+}
+
+func (store *memoryOfflineStore) getMemoryResourceTable(id ResourceID) (*memoryOfflineTable, error) {
 	table, has := store.tables[id]
 	if !has {
 		return nil, &TableNotFound{id.Name, id.Variant}
 	}
 	return table, nil
 }
 
+// Used to implement sort.Interface for sorting.
+type materializedRecords []ResourceRecord
+
+func (recs materializedRecords) Len() int {
+	return len(recs)
+}
+
+func (recs materializedRecords) Less(i, j int) bool {
+	return recs[i].Entity < recs[j].Entity
+}
+
+func (recs materializedRecords) Swap(i, j int) {
+	recs[i], recs[j] = recs[j], recs[i]
+}
+
+func (store *memoryOfflineStore) CreateMaterialization(id ResourceID) (Materialization, error) {
+	if id.Type != Feature {
+		return nil, errors.New("Only features can be materialized")
+	}
+	table, err := store.getMemoryResourceTable(id)
+	if err != nil {
+		return nil, err
+	}
+	matData := make(materializedRecords, 0, len(table.entityMap))
+	for _, records := range table.entityMap {
+		matRec := latestRecord(records)
+		matData = append(matData, matRec)
+	}
+	sort.Sort(matData)
+	matId := MaterializationID(uuid.NewString())
+	mat := &memoryMaterialization{
+		id:   matId,
+		data: matData,
+	}
+	store.materializations[matId] = mat
+	return mat, nil
+}
+
+type MaterializationNotFound struct {
+	id MaterializationID
+}
+
+func (err *MaterializationNotFound) Error() string {
+	return fmt.Sprintf("Materialization %s not found", err.id)
+}
+
+func (store *memoryOfflineStore) GetMaterialization(id MaterializationID) (Materialization, error) {
+	mat, has := store.materializations[id]
+	if !has {
+		return nil, &MaterializationNotFound{id}
+	}
+	return mat, nil
+}
+
+func latestRecord(recs []ResourceRecord) ResourceRecord {
+	latest := recs[0]
+	for _, rec := range recs {
+		if latest.TS.Before(rec.TS) {
+			latest = rec
+		}
+	}
+	return latest
+}
+
+func (store *memoryOfflineStore) CreateTrainingSet(def TrainingSetDef) error {
+	if err := def.check(); err != nil {
+		return err
+	}
+	label, err := store.getMemoryResourceTable(def.Label)
+	if err != nil {
+		return err
+	}
+	features := make([]*memoryOfflineTable, len(def.Features))
+	for i, id := range def.Features {
+		feature, err := store.getMemoryResourceTable(id)
+		if err != nil {
+			return err
+		}
+		features[i] = feature
+	}
+	labelRecs := label.records()
+	trainingData := make([]trainingRow, len(labelRecs))
+	for i, rec := range labelRecs {
+		featureVals := make([]interface{}, len(features))
+		for i, feature := range features {
+			featureVals[i] = feature.getLastValueBefore(rec.Entity, rec.TS)
```

Comment:
getlastvaluebefore returns nil if no entitymap to the given entity is set, but panics if it is but there are no resources. It should return nil in both cases I think

Reply:
Ill clarify the panic in a comment